XGenerate/src/main/java/com/xbreeze/xgenerate/model/ModelPreprocessor.java

```diff
@@ -24,242 +24,267 @@
+				for (int i = 0; i < result.getLength(); i++) {
+					Node node = result.item(i);
+					if (node.getNodeType() == Node.ELEMENT_NODE) {
+						node.getParentNode().removeChild(node);
+					} else if (node.getNodeType() == Node.ATTRIBUTE_NODE) {
+						((Attr)node).getOwnerElement().removeAttributeNode((Attr)node);
+					}
+				}	
+			} catch (XPathExpressionException e) {
+				throw new ModelPreprocessorException(String.format("Error processing XPATH expression for node removal %s, %s", mnr.getModelXPath(), e.getMessage()));
 			}
 		}
-		
 		// Return the modified XML document.
-		try {
-			return XMLUtils.getResultingXml(xm);
-		} catch (GeneratorException e) {
-			throw new ModelPreprocessorException(e);
-		}
+		return modelDoc;
 	}
-	
-	private static String performModelAttributeInjections(String preprocessedModel, ArrayList<ModelAttributeInjection> modelAttributeInjections, ArrayList<NamespaceConfig> namespaces) throws ModelPreprocessorException {
+
+	private static Document performModelAttributeInjections(Document modelDoc, 
+			ArrayList<ModelAttributeInjection> modelAttributeInjections, ArrayList<NamespaceConfig> namespaces)
+			throws ModelPreprocessorException {
 		logger.fine("Performing model attribute injections.");
-		
-		// Execute the model XPath on the Document.
-		VTDNav nv;
-		try {
-			// Set the VTD nav to namespace aware if the namespaces are defined.
-			nv = XMLUtils.getVTDNav(preprocessedModel, (namespaces != null && namespaces.size() > 0));
-		} catch (GeneratorException e) {
-			throw new ModelPreprocessorException(String.format("Error while reading model before performing model attribute injections: %s", e.getMessage()), e);
-		}
-		
-		// Create a modifier to modify the model document.
-		XMLModifier xm;
-		try {
-			xm = new XMLModifier(nv);
-		} catch (ModifyException e) {
-			throw new ModelPreprocessorException(e);
+
+		XPathFactoryImpl xPathfactory = new XPathFactoryImpl();
+		XPathEvaluator xpath = (XPathEvaluator) xPathfactory.newXPath();
+
+		// If namespaces are defined, create a namespace context object and set it on the xpath object.
+		if (namespaces != null && namespaces.size() > 0) {
+			ModelNamespaceContext nsContext = new ModelNamespaceContext(namespaces);
+			xpath.setNamespaceContext(nsContext);
 		}
 		
-
 		// Loop through the model attribute injections and process them.
 		for (ModelAttributeInjection mai : modelAttributeInjections) {
-			// Create an AutoPilot for querying the document.
-			AutoPilot ap = new AutoPilot(nv);
-			
-			// If namespaces are defined, register then on the auto pilot.
-			if (namespaces != null && namespaces.size() > 0) {
-				// Register the declared namespaces.
-				for (NamespaceConfig namespace : namespaces) {
-					ap.declareXPathNameSpace(namespace.getPrefix(), namespace.getNamespace());				
-				}
-			}
-			
 			try {
+				XPathExpressionImpl expr = (XPathExpressionImpl)xpath.compile(mai.getModelXPath());
```

Comment:
Is it an idea to refactor this compile method to an implemention in XmlUtils including making a singleton of the following piece of code:

```
		XPathFactoryImpl xPathfactory = new XPathFactoryImpl();
		XPathEvaluator xpath = (XPathEvaluator) xPathfactory.newXPath();
```

Or would we need a fresh XPathEvaluator every time?

Reply:
See reply to prev comment: this is now refactored to separate class and implemented as a singleton